### PR TITLE
Add top-k error

### PR DESCRIPTION
**Pull request by lzhao4ever**

### CODE REVIEW

File: paddle/parameter/Parameter.cpp [hunk 1]
```diff
@@ -375,10 +375,10 @@ bool Parameter::load(const std::string& filename) {
   std::ifstream fs(filename, std::ios_base::binary);
   if (!fs) {
     LOG(INFO) << "missing parameters [" << filename << "] while loading model.";
-    if (isStatic()) {
+    /*if (isStatic()) {
```
Comment (reyoung) [inline paddle/parameter/Parameter.cpp]: Why comment this lines? Either uncomment them or remove them.

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 2]
```diff
@@ -76,9 +83,56 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+
+        IVector::resizeOrCreate(
+            maxIds_, height * width, useGpu(arguments[0].deviceId));
+        Matrix::resizeOrCreate(
+            maxValues_, height, width, false, useGpu(arguments[0].deviceId));
+        output->rowMax(*maxIds_, *maxValues_);  // top-k values
+
+        int* ids = nullptr;
+        int* lbl = nullptr;
+        IVectorPtr dest = IVector::create(maxIds_->getSize(), false);
```
Comment (reyoung) [inline paddle/gserver/evaluators/Evaluator.cpp]:
```cpp
IVectorPtr dest = nullptr;
IVectorPtr dest2 = nullptr;
```

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 3]
```diff
@@ -76,9 +83,56 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+        IVectorPtr dest2 = IVector::create(label->getSize(), false);
+        if (useGpu(arguments[0].deviceId)) {
```
Comment (reyoung) [inline paddle/gserver/evaluators/Evaluator.cpp]:
```cpp
dest = IVector::create(label->getSize(), false);
dest2 = IVector::create(label->getSize(), false);
dest->copyFrom(*maxIds_);
dest2->copyFrom(*label);
```

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 4]
```diff
@@ -76,9 +83,56 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+
+          hl_memcpy_device2host((void*)dest2->getData(),
+                                (void*)label->getData(),
+                                sizeof(int) * label->getSize());
+          lbl = dest2->getData();
+        } else {
+          ids = maxIds_->getData();
+          lbl = label->getData();
+        }
+
```
Comment (reyoung) [inline paddle/gserver/evaluators/Evaluator.cpp]: errorMat2 is not necessary because we only need a counter(totalScore_) return.

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 5]
```diff
@@ -76,9 +83,56 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+        real* result2 = errorMat2->getData();
+        for (size_t i = 0; i < height; ++i) {
```
Comment (reyoung) [inline paddle/gserver/evaluators/Evaluator.cpp]:
```cpp
bool contain = false;
for (size_t j = 0; j<width && !contain; ++j) {
   error = (ids[i * width + j] == lbl[i]);  // label is in prediction ids 
}
if (!contain) {
    totalScore2_ += 1.0;
}
```

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 6]
```diff
@@ -76,9 +83,56 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+          result2[i] = (ids[i * width] != lbl[i]);  // initialize top-k error
+          for (size_t j = 1; j < width; ++j) {
+            if (result2[i] == 0.0) {
+              break;
+            }
+            result2[i] = (ids[i * width + j] != lbl[i]);  // top-k error
```
Comment (reyoung) [inline paddle/gserver/evaluators/Evaluator.cpp]: 这块的代码逻辑有点乱吧。

File: paddle/parameter/Parameter.cpp (hunk 1 above)
Comment (lzhao4ever) [inline paddle/parameter/Parameter.cpp]: It causes error when I tried to set static for some batch norm layers when their parameters are loaded from pre-trained model. I am trying to find out who added these lines and why.

File: paddle/gserver/evaluators/Evaluator.cpp (hunk 3 above)
Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: Rewrite these lines.

File: paddle/gserver/evaluators/Evaluator.cpp (hunk 4 above)
Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: True. Remove errorMat2.

File: paddle/gserver/evaluators/Evaluator.cpp (hunk 5 above)
Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: Variable "error" should be "contain".

File: paddle/gserver/evaluators/Evaluator.cpp (hunk 6 above)
Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: Rewrite this part.

File: paddle/gserver/evaluators/Evaluator.cpp [hunk 7]
```diff
@@ -76,9 +83,42 @@ class ClassificationErrorEvaluator : public Evaluator {
...
+    const MatrixPtr errorMat2 = Matrix::create(output->getHeight(),
+                                               1,
+                                               /* trans= */ false,
+                                               false);
+
     errorMat->zeroMem();
+
     if (label != nullptr) {
-      errorMat->classificationError(*output, *label);
+      errorMat->classificationError(*output, *label);  // top-1 error
```
Comment (gangliao) [inline paddle/gserver/evaluators/Evaluator.cpp]: 这里有点问题：
其实不需要计算完毕，再copy回来，更直接的方法是直接在CPU和GPU上做top k classification error.

具体做法：重载 Matrix.h
```c++	  
virtual void classificationError(MatrixPtr output, IVectorPtr label,
size_t topkSize = 1)
```

Comment (gangliao) [inline paddle/gserver/evaluators/Evaluator.cpp]: Matrix.cc

```c++
/*calulate the error of classification */
void GpuMatrix::classificationError(MatrixPtr output,
                                    IVectorPtr label,
                                    size_t topkSize) {
  size_t numSamples = this->getHeight();
  GpuMatrixPtr gpuOutput = std::dynamic_pointer_cast<GpuMatrix>(output);
  GpuIVectorPtr gpuLabel = std::dynamic_pointer_cast<GpuIVector>(label);
  GpuMatrixPtr gpuTopVal = std::make_shared<GpuMatrix>(numSamples, topkSize);
  GpuIVectorPtr gpuTopIds = std::make_shared<GpuIVector>(numSamples * topkSize);

  CHECK(gpuOutput && gpuLabel) << "Invalid argument pointer";
  CHECK(gpuTopVal && gpuTopIds) << "Allocate GPU memory failed";
  CHECK(gpuLabel->getSize() == numSamples) << "Vector size is not equal";
  CHECK(numSamples == gpuOutput->getHeight() && this->getWidth() == 1)
    << "Matrix dimensions are not equal";

  size_t dim = gpuOutput->getWidth();
  int ret = hl_matrix_classification_error(gpuTopVal->getData(),
                                           gpuTopVal->getStride(),
                                           gpuTopIds->getData(),
                                           gpuOutput->getData(),
                                           gpuOutput->getStride(),
                                           dim,
                                           topkSize,
                                           numSamples,
                                           gpuLabel->getData(),
                                           this->getData());

  CHECK_EQ(ret, HPPL_RET_OK) << "Error matrix classification ";
}
```

Comment (gangliao) [inline paddle/gserver/evaluators/Evaluator.cpp]: Matrix.cc classificationError on CPU
```c++
/*calulate the error of classification */
void GpuMatrix::classificationError(MatrixPtr output,
                                    IVectorPtr label,
                                    size_t topkSize) {
  size_t numSamples = this->getHeight();
  GpuMatrixPtr gpuOutput = std::dynamic_pointer_cast<GpuMatrix>(output);
  GpuIVectorPtr gpuLabel = std::dynamic_pointer_cast<GpuIVector>(label);
  GpuMatrixPtr gpuTopVal = std::make_shared<GpuMatrix>(numSamples, topkSize);
  GpuIVectorPtr gpuTopIds = std::make_shared<GpuIVector>(numSamples * topkSize);

  CHECK(gpuOutput && gpuLabel) << "Invalid argument pointer";
  CHECK(gpuTopVal && gpuTopIds) << "Allocate GPU memory failed";
  CHECK(gpuLabel->getSize() == numSamples) << "Vector size is not equal";
  CHECK(numSamples == gpuOutput->getHeight() && this->getWidth() == 1)
    << "Matrix dimensions are not equal";

  size_t dim = gpuOutput->getWidth();
  int ret = hl_matrix_classification_error(gpuTopVal->getData(),
                                           gpuTopVal->getStride(),
                                           gpuTopIds->getData(),
                                           gpuOutput->getData(),
                                           gpuOutput->getStride(),
                                           dim,
                                           topkSize,
                                           numSamples,
                                           gpuLabel->getData(),
                                           this->getData());

  CHECK_EQ(ret, HPPL_RET_OK) << "Error matrix classification ";
}
```

Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: 我想既计算top-1 error 又计算top-k error. 你的实现只给出了top-k error.

Comment (gangliao) [inline paddle/gserver/evaluators/Evaluator.cpp]: void classificationError(MatrixPtr output, IVectorPtr label,
size_t topkSize = 1)

默认就是top-1

Comment (qingqing01) [inline paddle/gserver/evaluators/Evaluator.cpp]: 廖刚的这种方式可以同时接两个Evaluator，一个是top-1，另一个是top-k。

Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: 知道了。谢谢。

Comment (lzhao4ever) [inline paddle/gserver/evaluators/Evaluator.cpp]: 将按建议修改。

**Comment by lzhao4ever** [comment]
谢谢提供详尽代码和说明：）
